File: scala/src/test/scala/nonamedb/test/specs/unit/storage/engines/MemoryEngineSpec.scala
package nonamedb.test.specs.unit.storage.engines

import akka.actor.ActorSystem
import akka.util.Timeout
import io.opentelemetry.api.trace.TracerProvider
import nonamedb.storage.engines.memory.MemoryEngine
import nonamedb.test.specs.unit.UnitSpec

import scala.concurrent.duration._

class MemoryEngineSpec extends UnitSpec with EngineBehavior {
  case class FixtureParam()

  private implicit val timeout: Timeout = 3.seconds
  private implicit val system: ActorSystem = ActorSystem("MemoryEngineSpec")

  private val testEngine = new MemoryEngine(TracerProvider.noop().get("MemoryEngineSpec"))

  "A MemoryEngine" should behave like basic(testEngine)
}

File: kotlin/src/test/kotlin/nonamedb/test/specs/unit/storage/engines/MemoryEngineSpec.kt
package nonamedb.test.specs.unit.storage.engines

import io.kotlintest.*
import io.kotlintest.specs.StringSpec
import nonamedb.storage.Done
import nonamedb.storage.engines.MemoryEngine

class MemoryEngineSpec : StringSpec(){
    init {
        val timeout = 5.seconds
        val testKey = "some key"
        val testValue = "some value".toByteArray()
        val updatedTestValue = "some updated value".toByteArray()
        val testEngine = MemoryEngine()

        "should fail to retrieve missing data" {
            val result = testEngine.get(testKey)
            eventually(timeout) {
                result.getCompleted() shouldBe null
            }
        }

        "should successfully add data" {
            val result = testEngine.put(testKey, testValue)
            eventually(timeout) {
                result.getCompleted() shouldBe Done
            }
        }

        "should successfully retrieve data" {
            val result = testEngine.get(testKey)
            eventually(timeout) {
                result.getCompleted() shouldBe testValue
            }
        }

        "should successfully update data" {
            val result = testEngine.put(testKey, updatedTestValue)
            eventually(timeout) {
                result.getCompleted() shouldBe Done
            }
        }

        "should successfully retrieve updated data" {
            val result = testEngine.get(testKey)
            eventually(timeout) {
                result.getCompleted() shouldBe updatedTestValue
            }
        }

        "should successfully remove data" {
            val result = testEngine.put(testKey, "".toByteArray())
            eventually(timeout) {
                result.getCompleted() shouldBe Done
            }
        }

        "should fail to retrieve removed data" {
            val result = testEngine.get(testKey)
            eventually(timeout) {
                result.getCompleted() shouldBe null
            }
        }
    }
}
